app/code/core/Mage/Adminhtml/controllers/CacheController.php

```diff
@@ -52,8 +52,12 @@ public function indexAction()
      */
     public function flushAllAction()
     {
-        Mage::dispatchEvent('adminhtml_cache_flush_all');
+        Mage::app()->getCacheInstance()->banUse('config');
+        Mage::getConfig()->reinit();
         Mage::app()->getCacheInstance()->flush();
```

Comment:
->reinit() also saves the cache, at least on line 325 of Mage_Core_Model_Config, why do we init+save then flush then save again? it is wanted?

Reply:
saveCache is called, but at that point in execution the useCache returns false so it has no effect. The reason to reinit first is to make sure that all of the latest data is loaded - for example if some xml files were updated or core_config_data was modified directly. It could be reinited after flush, but doing it before reduces the race condition time window between flush and saveCache.

Reply:
Updated so that now one button click can lock the cache from save, flush, apply updates, save, then unlock. I think this is much better than just flushing the cache since flushing leaves the apply updates to the next random web request - this ensures the user that clicked the button is the one that applies the updates while others wait. What do you think?

Note, this merges the latest version of #3530  - if that is merged first I'll rebase this PR.